### ISSUE TITLE
Replace `p1 - p2 < 1` with `p2 >= p1` everywhere

### ISSUE THREAD

**Issue by DemiMarie** [Mbed-TLS/mbedtls#6635]
### Suggested enhancement

Whenever mbedTLS does a check of the form `if (p1 - p2 < 1) { do something }`, the check should be rewritten as `if (p2 >= p1) { do something }`.

### Justification

Mbed TLS needs this because the second check has strictly less undefined behavior than the first.

**Comment by minosgalanakis** [comment]
Thank you for the proposal.

Could you elaborate on which edge cases are resulting in undefined behaviour, assuming that it is used for Natural positive numbers in the [0, N] space?

To my knowledge most of that type of checkes are in `size_t ` type of checks.

**Comment by tom-cosgrove-arm** [comment]
@DemiMarie How about raising this as a PR?

**Comment by DemiMarie** [comment]
@minosgalanakis:

> Thank you for the proposal.
> 
> Could you elaborate on which edge cases are resulting in undefined behaviour, assuming that it is used for Natural positive numbers in the [0, N] space?

To the best of my knowledge none of these can result in undefined behavior unless function preconditions are violated.  Otherwise I would have reported this as a potential security problem.  The undefined behavior appears when `p2 > p1` by a very large amount (more than `PTRDIFF_MAX`), which is a precondition violation in every case I looked at.
 
@tom-cosgrove-arm will do at some point.

**Comment by gilles-peskine-arm** [comment]
The reason the code has a lot of `end - p < 1` is that it uses the idiom `if (end - p < N) error();` before reading N bytes starting at `p`. I think using a different way to express the case with N=1 would have no advantage and would make the code less readable.

**Comment by DemiMarie** [comment]
@gilles-peskine-arm Fair point.  Perhaps there should be some sort of `end >= p` check at API entry points?  Also, I notice a lot of places where `end - p` is cast to `size_t`; is there a reason for this?

**Comment by gilles-peskine-arm** [comment]
> Perhaps there should be some sort of end >= p check at API entry points?

That might be a good idea. If `end` and `p` are in different objects, the behavior is undefined and in practice it's about an even chance of pass vs fail, but if `end` is a pointer to a segment of a bigger buffer, this catches the case where the cursor `p` has overrun the smaller buffer. We'd have to check the code size impact.

> Also, I notice a lot of places where end - p is cast to size_t; is there a reason for this?

In almost all of those, the result is compared with another value, and compilers would warn about a comparison between the signed value `end - p` and whatever unsigned thing is on the other side of the comparison operator. We know that `end - p` is non-negative from the cursor-before-end invariant, but the compiler doesn't know.

**Comment by DemiMarie** [comment]
> > Also, I notice a lot of places where end - p is cast to size_t; is there a reason for this?
> 
> In almost all of those, the result is compared with another value, and compilers would warn about a comparison between the signed value `end - p` and whatever unsigned thing is on the other side of the comparison operator. We know that `end - p` is non-negative from the cursor-before-end invariant, but the compiler doesn't know.

In a lot of cases the other side of the operator is actually a (signed) integer literal, but your point stands.
